announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/BaseAnnouncementService.java

```diff
@@ -1074,8 +1074,24 @@ public List<AnnouncementMessage> getMessages(String channelReference,Filter filt
 		catch (NullPointerException e) {
 			log.warn(e.getMessage());
 		}
-		return messageList;
 
+		String currentUserId = sessionManager.getCurrentSessionUserId();
+		return messageList.stream().filter(message -> {
+			List<String> selectedRoles = message.getProperties().getPropertyList("selectedRoles");
+			boolean isOwner = currentUserId.equals(message.getAnnouncementHeader().getFrom().getId());
+			if (selectedRoles == null || isOwner || securityService.isSuperUser()) {
+				return true;
+			} else {
+				String messageSiteId = entityManager.newReference(message.getReference()).getContext();
+				try {
+					Site msgSite = siteService.getSite(messageSiteId);
+					return selectedRoles.contains(msgSite.getMember(currentUserId).getRole().getId());
+				} catch (IdUnusedException idue) {
+					log.error("No site for id {}", messageSiteId);
+					return false;
+				}
+			}
+		}).collect(Collectors.toList());
```

Comment:
my instinct is that this is in the wrong place because this is just getMessages. what if a user has a direct link to an announcement? or if someone is moved from Teaching Assistant to Student and has an old notification.

i think we can just move your code here to the filter. please see a6f150be59dfcd532a0b11dd35473b66a8445862